_docs/deepfried_dd.md

```diff
@@ -0,0 +1,131 @@
+---
+layout: doc
+title: "Deepfried dd"
+learning_objectives:
+  - Working with files
+  - fseek() and fread()
+wikibook:
+  - "Files, Part 1: Working with files"
+---
+
+## Introduction
+
+`dd` is a command-line utility used to copy data to and from files. Since Linux treats many external devices (including USB drives) as files, this makes `dd` very powerful. For example, the tool can be used to create a backup image of your hard drive and store it as a file which can be uploaded to cloud storage. `dd` could also be used to directly clone one drive to another, write a bootable iso image to a USB drive, and much more. 
+
+For example, the following command would write an .img file (`if`) to a USB drive represented as /dev/disk4 (`of`), in chunks of 4 MB blocks (`bs`).
+
+```
+dd if=/path/to/bootable.img of=/dev/disk4 bs=4M
+```
+
+See the man pages for dd [here](http://man7.org/linux/man-pages/man1/dd.1.html), and example usage [here](https://linoxide.com/linux-command/linux-dd-command-create-1gb-file/). We suggest you get a feel of `dd` by using it to copy a file from one folder to another. Be careful, however! dd can easily be used to (accidentally or on purpose) to corrupt and entirely wipe disk drives and partitions, so make sure you know exactly what a `dd` command is going to do before running it! We recommend you make a testing folder on your VM and only use `dd` with paths pointing to that folder, so there's virtually no chance of overwriting something you don't want to.
+
+Note that Linux will automatically prevent you (usually) from writing to physical devices unless you run `dd` as root (i.e. via sudo).
+
+## Implementing dd
+For this assignment, you will be implementing the `dd` utility in C. Your `dd` implementation will copy data from an input file to an output file in a manner specified by its arguments.
+
+### Background: Blocks
+
+Though "blocks" are a fundamental filesystem concept, this assignment does not rely on heavy filesystems knowledge. For the purposes of this assignment, a *block* is simply a single unit of data. For example, if we run `./dd -i input_file -o output_file -b 128`, we are telling `dd` to copy `input_file` to `output_file`, 128 bytes at a time. Your code should write the first 128 bytes of `input_file` to `output_file`, then the next 128, and so on and so forth in a loop until the `input_file` is exhausted (as opposed to copying the entire file in one go). The arguments of `dd` along with more example usage can be found below.
```

Comment:
Hmm, I tried reading this pretending I had no background knowledge on blocks. While the example is great and should stay, I don't think this effectively explains blocks.

Reply:
Updated in the the latest commit.